File: backend/app/schemas/security_schemas.py
from typing import List, Literal

from pydantic import BaseModel

from app.schemas.analysis_details_schemas import AnalysisDetails


class PromptCheckRequest(BaseModel):
    prompt: str


class WebpageCheckRequest(BaseModel):
    visible_text: str
    hidden_text: str
    html_comments: str
    meta_tags: str
    input_values: str
    page_title: str
    url: str
    aria_text: str = ""
    iframe_content: str = ""
    shadow_dom_content: str = ""
    external_javascript: str = ""
    inline_javascript: str = ""
    css_content: str = ""
    css_generated_content: str = ""
    network_responses: str = ""
    websocket_messages: str = ""
    service_worker_activity: str = ""
    source_maps: str = ""
    redirects: str = ""
    third_party_resources: str = ""
    suspicious_domains: str = ""
    frame_navigation: str = ""
    runtime_script_activity: str = ""
    loaded_resources: str = ""
    dom_snapshot_content: str = ""


class SecurityCheckResponse(BaseModel):
    allowed: bool
    label: Literal["benign", "malicious"]
    confidence: float
    risk_level: Literal["low", "medium", "high"]
    summary_reason: str
    matched_patterns: List[str]
    source: Literal["direct_prompt", "webpage_content"]
    timestamp: str
    analysis_details: AnalysisDetails

File: backend-node/src/schemas/security.schemas.ts
import { Type, type Static } from '@sinclair/typebox'
import { AnalysisDetailsSchema } from './analysisDetails.schemas.js'

export const PromptCheckRequestSchema = Type.Object({
  prompt: Type.String(),
})

// Field-for-field port of WebpageCheckRequest. Required fields intentionally
// match Pydantic's defaults: only the original seven core capture fields are
// required, while the newer channels default to an empty string.
export const WebpageCheckRequestSchema = Type.Object({
  visible_text: Type.String(),
  hidden_text: Type.String(),
  html_comments: Type.String(),
  meta_tags: Type.String(),
  input_values: Type.String(),
  page_title: Type.String(),
  url: Type.String(),
  aria_text: Type.Optional(Type.String()),
  iframe_content: Type.Optional(Type.String()),
  shadow_dom_content: Type.Optional(Type.String()),
  external_javascript: Type.Optional(Type.String()),
  inline_javascript: Type.Optional(Type.String()),
  css_content: Type.Optional(Type.String()),
  css_generated_content: Type.Optional(Type.String()),
  network_responses: Type.Optional(Type.String()),
  websocket_messages: Type.Optional(Type.String()),
  service_worker_activity: Type.Optional(Type.String()),
  source_maps: Type.Optional(Type.String()),
  redirects: Type.Optional(Type.String()),
  third_party_resources: Type.Optional(Type.String()),
  suspicious_domains: Type.Optional(Type.String()),
  frame_navigation: Type.Optional(Type.String()),
  runtime_script_activity: Type.Optional(Type.String()),
  loaded_resources: Type.Optional(Type.String()),
  dom_snapshot_content: Type.Optional(Type.String()),
})

export const SecurityCheckResponseSchema = Type.Object({
  allowed: Type.Boolean(),
  label: Type.Union([Type.Literal('benign'), Type.Literal('malicious')]),
  confidence: Type.Number(),
  risk_level: Type.Union([Type.Literal('low'), Type.Literal('medium'), Type.Literal('high')]),
  summary_reason: Type.String(),
  matched_patterns: Type.Array(Type.String()),
  source: Type.Union([Type.Literal('direct_prompt'), Type.Literal('webpage_content')]),
  timestamp: Type.String(),
  analysis_details: AnalysisDetailsSchema,
})

export type PromptCheckRequest = Static<typeof PromptCheckRequestSchema>
export type WebpageCheckRequest = Static<typeof WebpageCheckRequestSchema>
export type SecurityCheckResponse = Static<typeof SecurityCheckResponseSchema>
